### PR TITLE
limit retries of imagesignature import and don't log failures

### PR DESCRIPTION
**Pull request by bparees**
fixes https://github.com/openshift/origin/issues/17172

### CODE REVIEW

**Comment by smarterclayton** [comment]
Nov  4 23:42:39.194: INFO: net/http: TLS handshake timeout
[AfterEach] [Feature:DeploymentConfig] deploymentconfigs
  /tmp/openshift/build-rpm-release/tito/rpmbuild-

Seems very wrong

**Comment by bparees** [comment]
> Seems very wrong

but unrelated to these changes, no?  should be opened as an issue?

**Comment by smarterclayton** [comment]
We have a bevy of networking related flakes, some which might actually be the master, some which might be node host networking @openshift/sig-networking @openshift/master

**Comment by smarterclayton** [comment]
@openshift/sig-master

File: pkg/image/controller/signature/signature_import_controller.go [hunk 1]
```diff
@@ -107,7 +107,9 @@ func (s *SignatureImportController) work() bool {
 	err := s.syncImageSignatures(key.(string))
 	if err != nil {
 		utilruntime.HandleError(fmt.Errorf("error syncing image %s, it will be retried: %v", key.(string), err))
-		s.queue.AddRateLimited(key)
+		if s.queue.NumRequeues(key) < 5 {
```
Comment (smarterclayton) [inline pkg/image/controller/signature/signature_import_controller.go]: Pull the magic number out

File: pkg/image/controller/signature/container_image_downloader.go [hunk 2]
```diff
@@ -46,7 +46,8 @@ func (s *containerImageSignatureDownloader) DownloadImageSignatures(image *image
 
 	signatures, err := source.GetSignatures(ctx)
 	if err != nil {
-		return nil, err
+		glog.V(4).Infof("Failed to get signatures for %v due to: %v", source.Reference(), err)
```
Comment (smarterclayton) [inline pkg/image/controller/signature/container_image_downloader.go]: Hrm, can we make this an ignorable error instead?  I don’t want to make it too complex, but would prefer to eat the error as late as possible.

Another option is a rate limiter on this type of error (if rate exceeded, don’t log).  Maybe burst 100 rate 1/10s or something.

File: pkg/image/controller/signature/signature_import_controller.go (hunk 1 above)
Comment (bparees) [reply pkg/image/controller/signature/signature_import_controller.go]: pull it out to where?  my goal here was to say "hey, if we can't retrieve this after a couple tries, it's not important enough to keep retrying" (but we'll eventually retry on relist of course).

File: pkg/image/controller/signature/container_image_downloader.go (hunk 2 above)
Comment (bparees) [reply pkg/image/controller/signature/container_image_downloader.go]: other than whatever handleerror does, do we have a utility for managing rate limits of logging/errors?

**Comment by bparees** [comment]
@smarterclayton i've reworked the error logic to eat the error later (did not introduce rate limiting and don't intend to unless we have a generic utility for it already).  Still not sure what you want to see done with the magic retry number.

File: pkg/image/controller/signature/container_image_downloader.go [hunk 3]
```diff
@@ -26,6 +26,10 @@ func NewContainerImageSignatureDownloader(ctx context.Context, timeout time.Dura
 	}
 }
 
+type GetSignaturesError struct {
```
Comment (mfojtik) [inline pkg/image/controller/signature/container_image_downloader.go]: deserves godoc

File: pkg/image/controller/signature/signature_import_controller.go [hunk 4]
```diff
@@ -106,8 +106,13 @@ func (s *SignatureImportController) work() bool {
 
 	err := s.syncImageSignatures(key.(string))
 	if err != nil {
-		utilruntime.HandleError(fmt.Errorf("error syncing image %s, it will be retried: %v", key.(string), err))
-		s.queue.AddRateLimited(key)
+		if _, ok := err.(GetSignaturesError); !ok {
+			utilruntime.HandleError(fmt.Errorf("error syncing image %s, it will be retried: %v", key.(string), err))
+		}
+
+		if s.queue.NumRequeues(key) < 5 {
```
Comment (mfojtik) [inline pkg/image/controller/signature/signature_import_controller.go]: can you make the 5 a const? `maxSignatureFetchRetries` or something

File: pkg/image/controller/signature/signature_import_controller.go [hunk 5]
```diff
@@ -106,8 +106,13 @@ func (s *SignatureImportController) work() bool {
 
 	err := s.syncImageSignatures(key.(string))
 	if err != nil {
-		utilruntime.HandleError(fmt.Errorf("error syncing image %s, it will be retried: %v", key.(string), err))
-		s.queue.AddRateLimited(key)
+		if _, ok := err.(GetSignaturesError); !ok {
+			utilruntime.HandleError(fmt.Errorf("error syncing image %s, it will be retried: %v", key.(string), err))
```
Comment (mfojtik) [inline pkg/image/controller/signature/signature_import_controller.go]: can we log the error under V(4) or V(5)? Can help debugging.

File: pkg/image/controller/signature/container_image_downloader.go [hunk 6]
```diff
@@ -46,7 +50,8 @@ func (s *containerImageSignatureDownloader) DownloadImageSignatures(image *image
 
 	signatures, err := source.GetSignatures(ctx)
 	if err != nil {
-		return nil, err
+		glog.V(4).Infof("Failed to get signatures for %v due to: %v", source.Reference(), err)
+		return []imageapi.ImageSignature{}, GetSignaturesError{err}
```
Comment (mfojtik) [inline pkg/image/controller/signature/container_image_downloader.go]: just return nil?

File: pkg/image/controller/signature/container_image_downloader.go (hunk 3 above)
Comment (bparees) [reply pkg/image/controller/signature/container_image_downloader.go]: no it doesn't.

File: pkg/image/controller/signature/signature_import_controller.go (hunk 4 above)
Comment (bparees) [reply pkg/image/controller/signature/signature_import_controller.go]: why? we're not using it anywhere else.  if you're not using a constant in two places, there's not much point in making it a constant.

File: pkg/image/controller/signature/signature_import_controller.go (hunk 5 above)
Comment (bparees) [reply pkg/image/controller/signature/signature_import_controller.go]: it was already logged when we got the error.

Comment (bparees) [reply pkg/image/controller/signature/signature_import_controller.go]: glog.V(4).Infof("Failed to get signatures for %v due to: %v", source.Reference(), err)
		return []imageapi.ImageSignature{}, GetSignaturesError{err}

File: pkg/image/controller/signature/container_image_downloader.go (hunk 6 above)
Comment (bparees) [reply pkg/image/controller/signature/container_image_downloader.go]: I followed the pattern you established in this function.

File: pkg/image/controller/signature/container_image_downloader.go (hunk 3 above)
Comment (bparees) [reply pkg/image/controller/signature/container_image_downloader.go]: (what would it say?  "GetSignaturesError is an error returned when GetSignatures encounters an error"?)

File: pkg/image/controller/signature/signature_import_controller.go (hunk 4 above)
Comment (tnozicka) [reply pkg/image/controller/signature/signature_import_controller.go]: I would say there is still point in that: if someone writing a new peace of code will be looking for such constant in the package and the is none he will either create it or use your argument about not using a constant if he uses it only in one place and magic numbers will spread

File: pkg/image/controller/signature/container_image_downloader.go (hunk 2 above)
Comment (smarterclayton) [reply pkg/image/controller/signature/container_image_downloader.go]: we have a rate limiter class - it's fairly easy to add.  Can be omitted here.

File: pkg/image/controller/signature/signature_import_controller.go (hunk 4 above)
Comment (smarterclayton) [reply pkg/image/controller/signature/signature_import_controller.go]: If you have constants they also spread.  Private constants in a package are generally what we try to do though to pull them into one place and put documentation on them.

**Comment by smarterclayton** [comment]
This looks fine to me

**Pull request merged**